File: src/floww/core/workspace/backends/niri.py
import json
import logging
import subprocess

from floww.utils import run_command

from .base import WorkspaceBackend

logger = logging.getLogger(__name__)


class NiriBackend(WorkspaceBackend):
    """Niri via ``niri msg`` (dynamic workspaces on the focused output)."""

    def __init__(self, niri_cmd: str = "niri"):
        self._niri_cmd = niri_cmd

    @staticmethod
    def _workflow_target_to_idx(target: int) -> int:
        """
        Floww YAML ``target`` and niri ``idx`` / ``focus-workspace`` numeric refs are **1-based**
        (first workspace is ``1``). Values below ``1`` are clamped to ``1``.
        """
        return target if target >= 1 else 1

    def _fetch_workspaces_json(self) -> list | None:
        try:
            cmd = [self._niri_cmd, "msg", "-j", "workspaces"]
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                check=True,
            )
            data = json.loads(result.stdout)
            return data if isinstance(data, list) else None
        except Exception as e:
            logger.debug("niri workspaces fetch failed: %s", e)
            return None

    @staticmethod
    def _focused_workspace_idx(workspaces: list) -> int | None:
        for ws in workspaces:
            if not isinstance(ws, dict):
                continue
            if ws.get("is_focused") or ws.get("is_active"):
                idx = ws.get("idx")
                if isinstance(idx, int):
                    return idx
        return None

    def _is_already_focused_on_target(self, target: int) -> bool:
        """
        True if the focused workspace's **1-based** ``idx`` equals this workflow ``target``.

        Used to avoid ``focus-workspace`` when already there (e.g. ``workspace-auto-back-and-forth``).
        False if state cannot be determined.
        """
        workspaces = self._fetch_workspaces_json()
        if workspaces is None:
            return False
        desired = self._workflow_target_to_idx(target)
        cur = self._focused_workspace_idx(workspaces)
        return cur is not None and cur == desired

    def switch(self, target: int) -> bool:
        """Focus workspace ``target`` (1-based, same convention as niri ``idx``)."""
        try:
            if self._is_already_focused_on_target(target):
                logger.debug(
                    "niri: already on workspace idx %s; skipping focus-workspace",
                    self._workflow_target_to_idx(target),
                )
                return True

            idx = self._workflow_target_to_idx(target)
            cmd = [
                self._niri_cmd,
                "msg",
                "action",
                "focus-workspace",
                str(idx),
            ]
            return run_command(cmd)
        except Exception as e:
            logger.error("niri focus-workspace failed: %s", e)
            return False

    def get_total_workspaces(self) -> int:
        workspaces = self._fetch_workspaces_json()
        if workspaces is None:
            return 0
        if not workspaces:
            return 1
        return len(workspaces)

File: internal/workspace/backends/niri.go
package backends

import (
	"encoding/json"
	"log/slog"
	"strconv"
)

// NiriBackend controls Niri workspaces via “niri msg“.
type NiriBackend struct {
	niriCmd string
}

// NewNiriBackend returns a NiriBackend that uses the given niri command
// (default "niri").
func NewNiriBackend() *NiriBackend {
	return &NiriBackend{
		niriCmd: "niri",
	}
}

// niriWorkspace represents a single workspace in niri's JSON output.
type niriWorkspace struct {
	ID        int  `json:"id"`
	Idx       int  `json:"idx"`
	IsFocused bool `json:"is_focused"`
	IsActive  bool `json:"is_active"`
}

// workflowTargetToIdx converts a workflow target to niri's 1-based idx,
// clamping values below 1 to 1.
func workflowTargetToIdx(target int) int {
	if target < 1 {
		return 1
	}
	return target
}

// fetchWorkspacesJSON runs "niri msg -j workspaces" and returns the parsed
// list of workspaces.
func (b *NiriBackend) fetchWorkspacesJSON() ([]niriWorkspace, error) {
	cmd := execCmd(b.niriCmd, "msg", "-j", "workspaces")
	out, err := cmd.Output()
	if err != nil {
		slog.Debug("niri workspaces fetch failed", "error", err)
		return nil, err
	}

	var workspaces []niriWorkspace
	if err := json.Unmarshal(out, &workspaces); err != nil {
		slog.Debug("niri workspaces parse failed", "error", err)
		return nil, err
	}
	return workspaces, nil
}

// focusedWorkspaceIdx returns the idx of the focused (or active) workspace.
func focusedWorkspaceIdx(workspaces []niriWorkspace) (int, bool) {
	for _, ws := range workspaces {
		if ws.IsFocused || ws.IsActive {
			return ws.Idx, true
		}
	}
	return 0, false
}

// isAlreadyFocusedOnTarget checks whether the currently focused workspace
// has the same idx as the desired target. Returns false when the workspace
// state cannot be determined, causing Switch to fall through to
// focus-workspace.
func (b *NiriBackend) isAlreadyFocusedOnTarget(target int) bool {
	workspaces, err := b.fetchWorkspacesJSON()
	if err != nil {
		return false
	}
	desired := workflowTargetToIdx(target)
	cur, ok := focusedWorkspaceIdx(workspaces)
	if !ok {
		return false
	}
	return cur == desired
}

// Switch focuses the given workspace (1-based). It skips the
// focus-workspace action when already on the target workspace
// (auto-back-and-forth guard).
func (b *NiriBackend) Switch(target int) bool {
	if b.isAlreadyFocusedOnTarget(target) {
		slog.Debug("niri: already on workspace", "idx", workflowTargetToIdx(target))
		return true
	}

	idx := workflowTargetToIdx(target)
	return runCmd(b.niriCmd, "msg", "action", "focus-workspace", strconv.Itoa(idx))
}

// GetTotalWorkspaces returns the number of niri workspaces on the focused
// output. Returns 1 when the list is empty (niri always has at least one
// workspace) and 0 when the query fails.
func (b *NiriBackend) GetTotalWorkspaces() int {
	workspaces, err := b.fetchWorkspacesJSON()
	if err != nil {
		return 0
	}
	if len(workspaces) == 0 {
		return 1
	}
	return len(workspaces)
}

// GetAppendBaseOffset returns max(0, total-1), the default offset for
// append mode.
func (b *NiriBackend) GetAppendBaseOffset() int {
	return AppendBaseOffset(b.GetTotalWorkspaces())
}
